src/herder/test/HerderTests.cpp

```diff
@@ -5339,3 +5340,199 @@ TEST_CASE("exclude transactions by operation type", "[herder]")
                 TransactionQueue::AddResult::ADD_STATUS_PENDING);
     }
 }
+
+TEST_CASE("delay sending DONT_HAVE", "[herder]")
+{
+    VirtualClock clock;
+    auto const numNodes = 2;
+    std::vector<std::shared_ptr<Application>> apps;
+    std::chrono::milliseconds const epsilon{1};
+    auto s = SecretKey::pseudoRandomForTesting();
+
+    for (auto i = 0; i < numNodes; i++)
+    {
+        Config cfg = getTestConfig(i);
+        cfg.FLOOD_DEMAND_BACKOFF_DELAY_MS = std::chrono::milliseconds(200);
+        cfg.FLOOD_DEMAND_PERIOD_MS = std::chrono::milliseconds(200);
+        // Using a small tq set size such as 50 may lead to an unexpectedly
+        // small advert/demand size limit.
+        cfg.MANUAL_CLOSE = false;
+        cfg.TESTING_UPGRADE_MAX_TX_SET_SIZE = 1000;
+        cfg.SEND_DONT_HAVE_DELAY = std::chrono::milliseconds{300};
+        // Setting validators
+        cfg.QUORUM_SET.validators.emplace_back(s.getPublicKey());
+        // cfg.LEDGER_PROTOCOL_VERSION
+        apps.push_back(createTestApplication(clock, cfg));
+    }
+
+    auto connection =
+        std::make_shared<LoopbackPeerConnection>(*apps[0], *apps[1]);
+    testutil::crankFor(clock, std::chrono::seconds(5));
+    REQUIRE(connection->getInitiator()->isAuthenticated());
+    REQUIRE(connection->getAcceptor()->isAuthenticated());
+
+    // TODO: maybe replace these lambdas with static definitions?
+    auto createTxn = [](auto n) {
+        StellarMessage txn;
+        txn.type(TRANSACTION);
+        Memo memo(MEMO_TEXT);
+        memo.text() = "tx" + std::to_string(n);
+        txn.transaction().v0().tx.memo = memo;
+
+        return std::make_shared<StellarMessage>(txn);
+    };
+
+    auto createGetTxSetMessage = [](uint256 const& setID) {
+        StellarMessage getTxSet;
+        getTxSet.type(GET_TX_SET);
+        getTxSet.txSetHash() = setID;
+
+        return std::make_shared<StellarMessage>(getTxSet);
+    };
+
+    auto createTxSetMessage = [](TxSetFrameConstPtr txSet) {
+        StellarMessage msg;
+        msg.type(TX_SET);
```

Comment:
Since we recently enabled a new protocol (v20) in master which this change was rebased on top of, this should be generalized tx set.